# How to use danda
The first step you need to do is to collect analysis, study the report, then decide if you want to go with the automatic danda improvements

In [1]:
#from pprint import pprint
#import numpy as np
import pandas as pd
# notice the need to danda import to have accessor df
import danda # noqa: F401  # registers the pandas accessor
from danda.report_renderer import ReportRenderer

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv ")
df

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa
...,...,...,...,...,...
145,6.7,3.0,5.2,2.3,virginica
146,6.3,2.5,5.0,1.9,virginica
147,6.5,3.0,5.2,2.0,virginica
148,6.2,3.4,5.4,2.3,virginica


In [3]:
report = df.dg.clean().dg.optimize().dg.analyze().dg.report
renderer = ReportRenderer()
print(renderer.render(report))
#print(json.dumps(report, indent=4, default=str))

Danda Report

Clean
-----

EmptyRowsPlugin
    Number of deleted rows: 0

EmptyColumnsPlugin
    Number of deleted columns: 0

DropDuplicates
    Number of deleted rows: 1

Execution
---------
Plugins executed : 7
Plugin names     : EmptySpacesPlugin, NormalizeMissingValuesPlugin, SparseColumnsPlugin, SparseRowsPlugin, EmptyRowsPlugin, EmptyColumnsPlugin, DropDuplicates

Optimize
--------

BooleanTypePlugin
    no columns converted

CategoryTypePlugin
    Converted the following columns to category: species

Execution
---------
Plugins executed : 4
Plugin names     : BooleanTypePlugin, DateTimeTypePlugin, NumericTypePlugin, CategoryTypePlugin
Memory before    : 8,412 bytes
Memory after     : 6,159 bytes
Memory saved     : 2,253 bytes (26.8%)

Analyze
-------

ColumnSummary
    Column Summary:
    
    sepal_length
    Type: float64
    Missing: 0
    Unique: 35
    
    sepal_width
    Type: float64
    Missing: 0
    Unique: 23
    
    petal_length
    Type: float64
    Missing: 0
  

Looking at the report we can see:
* Number of deleted rows: 1
* Converted the following columns to category: species
* Memory saved     : 2,253 bytes (26.8%)

Let us see if there is any thing else we can improve



In [5]:
new_df = df.dg.clean().dg.optimize()


## Let us verify

In [6]:
print(df.info())
print(new_df.info())

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 118.9 KB
None
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int64   
 1   Survived     891 non-null